Description: Update the Accountsservice user background
 Accountsservice outside ubuntu didn't have the wallpaper/background setting,
 now it does. Support new AccountsService method of setting background file.
Author: Lionel Le Folgoc <mrpouit@ubuntu.com>
Forwarded: not-needed
Last-Update: 2018-02-17

---
 src/xfce-desktop.c |  104 ++++++++++++++++++++++++++++++++++++++++++++++++++++-
 1 file changed, 103 insertions(+), 1 deletion(-)

--- a/src/xfce-desktop.c
+++ b/src/xfce-desktop.c
@@ -63,6 +63,7 @@
 #include <glib.h>
 #include <gdk/gdkx.h>
 #include <gtk/gtk.h>
+#include <gio/gio.h>
 
 #ifdef ENABLE_DESKTOP_ICONS
 #include "xfdesktop-icon-view.h"
@@ -119,6 +120,8 @@
     GtkWidget *icon_view;
     gdouble system_font_size;
 #endif
+
+    gchar *last_filename;
 };
 
 enum
@@ -366,11 +369,93 @@
 }
 
 static void
+set_accountsservice_user_bg(const gchar *background)
+{
+    GDBusConnection *bus;
+    GVariant *variant;
+    GError *error = NULL;
+    gchar *object_path = NULL;
+
+    bus = g_bus_get_sync (G_BUS_TYPE_SYSTEM, NULL, &error);
+    if (bus == NULL) {
+        g_warning ("Failed to get system bus: %s", error->message);
+        g_error_free (error);
+        return;
+    }
+
+    variant = g_dbus_connection_call_sync (bus,
+                                           "org.freedesktop.Accounts",
+                                           "/org/freedesktop/Accounts",
+                                           "org.freedesktop.Accounts",
+                                           "FindUserByName",
+                                           g_variant_new ("(s)", g_get_user_name ()),
+                                           G_VARIANT_TYPE ("(o)"),
+                                           G_DBUS_CALL_FLAGS_NONE,
+                                           -1,
+                                           NULL,
+                                           &error);
+
+    if (variant == NULL) {
+        DBG("Could not contact accounts service to look up '%s': %s",
+            g_get_user_name (), error->message);
+        g_error_free(error);
+        g_object_unref (bus);
+        return;
+    }
+
+    g_variant_get(variant, "(o)", &object_path);
+    g_variant_unref(variant);
+
+    variant = g_dbus_connection_call_sync (bus,
+                                           "org.freedesktop.Accounts",
+                                           object_path,
+                                           "org.freedesktop.DBus.Properties",
+                                           "Set",
+                                           g_variant_new ("(ssv)",
+                                                          "org.freedesktop.DisplayManager.AccountsService",
+                                                          "BackgroundFile",
+                                                          g_variant_new_string (background ? background : "")),
+                                           G_VARIANT_TYPE ("()"),
+                                           G_DBUS_CALL_FLAGS_NONE,
+                                           -1,
+                                           NULL,
+                                           &error);
+    if (variant != NULL) {
+        g_variant_unref (variant);
+    } else {
+        g_warning ("Failed to set the background '%s': %s", background, error->message);
+        g_clear_error (&error);
+    }
+
+    /* Also attempt the old method (patch not upstreamed into AccountsService */
+    variant = g_dbus_connection_call_sync (bus,
+                                           "org.freedesktop.Accounts",
+                                           object_path,
+                                           "org.freedesktop.Accounts.User",
+                                           "SetBackgroundFile",
+                                           g_variant_new ("(s)", background ? background : ""),
+                                           G_VARIANT_TYPE ("()"),
+                                           G_DBUS_CALL_FLAGS_NONE,
+                                           -1,
+                                           NULL,
+                                           &error);
+    if (variant != NULL) {
+        g_variant_unref (variant);
+    } else {
+        g_warning ("Failed to set the background '%s': %s", background, error->message);
+        g_clear_error (&error);
+    }
+
+    g_object_unref (bus);
+}
+
+static void
 backdrop_changed_cb(XfceBackdrop *backdrop, gpointer user_data)
 {
     XfceDesktop *desktop = XFCE_DESKTOP(user_data);
     GdkPixmap *pmap = desktop->priv->bg_pixmap;
     GdkScreen *gscreen = desktop->priv->gscreen;
+    gchar *new_filename = NULL;
     GdkRectangle rect;
     GdkRegion *clip_region = NULL;
     gint i, monitor = -1, current_workspace;
@@ -401,6 +486,20 @@
     }
     if(monitor == -1)
         return;
+    /* notify Accountsservice of the new bg (only for monitor0) */
+    if(monitor == 0)
+    {
+        if (xfce_desktop_get_current_workspace(desktop) == 0)
+        {
+            new_filename = g_strdup(xfce_backdrop_get_image_filename(backdrop));
+            if (g_strcmp0(desktop->priv->last_filename, new_filename) != 0)
+            {
+                desktop->priv->last_filename = g_strdup(new_filename);
+                set_accountsservice_user_bg(xfce_backdrop_get_image_filename(backdrop));
+            }
+            g_free(new_filename);
+        }
+    }
 
 #ifdef G_ENABLE_DEBUG
     monitor_name = gdk_screen_get_monitor_plug_name(gscreen, monitor);
@@ -1449,6 +1548,8 @@
 
     xfce_desktop_connect_settings(desktop);
     
+    desktop->priv->last_filename = g_strdup("");
+
     return GTK_WIDGET(desktop);
 }
 
